modulemd_tools/modulemd_tools/yaml.py

```diff
@@ -0,0 +1,317 @@
+            for depstream in old_deps.get_buildtime_streams(depname):
+                new_deps.add_buildtime_stream(depname, depstream)
+
+    if requires or buildrequires:
+        mod_stream.clear_dependencies()
+        mod_stream.add_dependencies(new_deps)
+
+    idx2 = Modulemd.ModuleIndex.new()
+    idx2.add_module_stream(mod_stream)
+    return idx2.dump_to_string()
+
+
+def upgrade(mod_yaml, version):
+    """
+    Upgrade the input module YAML from an older version to a newer one.
+    Downgrades aren't supported even in case where it would be possible.
+    """
+    parsed = yaml.safe_load(mod_yaml or "")
+    if not parsed or not "version" in parsed:
+        raise ValueError("Missing modulemd version")
+
+    supported = [1, 2]
+    if parsed["version"] not in supported or version not in supported:
+        raise ValueError("Unexpected modulemd version")
+
+    if parsed["version"] > version:
+        raise ValueError("Cannot downgrade modulemd version")
+
+    mod_stream = Modulemd.ModuleStream.new(parsed["version"])
+    mod_stream = mod_stream.read_string(mod_yaml, True,
+                                        parsed["data"].get("name", ""),
+                                        parsed["data"].get("stream", ""))
+    mod_stream_upgraded = mod_stream.upgrade(version)
+    return _stream2yaml(mod_stream_upgraded)
+
+
+def load(path):
+    """
+    Load modulemd YAML definition from a file
+    """
+    with open(path, "r") as yaml_file:
+        mod_yaml = yaml_file.read()
+    validate(mod_yaml)
+    return mod_yaml
+
+
+def dump(mod_yaml, dest=None):
+    """
+    Dump modulemd YAML definition into a file
+
+    If `dest` is not specified, the file will be created in the current working
+    directory and it's name is going to be generated from the module attributes
+    in the `N:S:V:C:A.modulemd.yaml` format.
+
+    If `dest` is a directory, then the file is going to be generated in that
+    directory.
+
+    If `dest` is pointing to a (nonexisting) file, it is going to be dumped in
+    its place.
+    """
+    filename = _generate_filename(mod_yaml)
+
+    path = dest
+    if not path:
+        path = os.path.join(os.getcwd(), filename)
+    elif os.path.isdir(path):
+        path = os.path.join(dest, filename)
+
+    with open(path, "w") as moduleyaml:
+        moduleyaml.write(mod_yaml)
+
+
+def _generate_filename(mod_yaml):
+    """
+    Generate filename for a module yaml
+    """
+    mod_stream = _yaml2stream(mod_yaml)
+    return "{N}:{S}:{V}:{C}:{A}.modulemd.yaml".format(
+        N=mod_stream.get_module_name(),
+        S=mod_stream.get_stream_name(),
+        V=mod_stream.get_version(),
+        C=mod_stream.get_context(),
+        A=mod_stream.get_arch() or "noarch")
+
+
+def _yaml2stream(mod_yaml):
+    if not mod_yaml:
+        raise ValueError("Provided YAML did not begin with a module document")
+
+    idx = Modulemd.ModuleIndex.new()
+    idx.update_from_string(mod_yaml, strict=True)
```

Comment:
Be careful about hard-coding `strict=True`. We usually only want to be strict in the tools that are creating repodata or building modules. We don't want to be strict in the tools that *consume* the metadata, because we can have problems when the repodata is created by a newer version of libmodulemd than the one being used by the client (e.g. DNF).